File: src/Starter/Controls/KeyboardShortcutInput.axaml.cs
using Avalonia;
using Avalonia.Controls;
using Avalonia.Input;
using Avalonia.Interactivity;
using Starter.ViewModels;

namespace Starter.Controls;

public partial class KeyboardShortcutInput : UserControl
{
    public static readonly StyledProperty<Control?> ControlToFocusProperty = AvaloniaProperty.Register<KeyboardShortcutInput, Control?>(nameof(ControlToFocus));

    public Control? ControlToFocus
    {
        get => GetValue(ControlToFocusProperty);
        set => SetValue(ControlToFocusProperty, value);
    }

    public KeyboardShortcutInput() => InitializeComponent();

    protected override void OnDataContextChanged(EventArgs e)
    {
        base.OnDataContextChanged(e);
        if (DataContext is not KeyboardShortcutInputViewModel vm) return;
        vm.StoppedListening += OnStoppedListening;
    }

    private void OnStoppedListening() => ControlToFocus?.Focus();

    private void TextBoxKeyDown(object? sender, KeyEventArgs e)
    {
        if (DataContext is not KeyboardShortcutInputViewModel vm) return;
        vm.OnKeyDown(e.Key);
        e.Handled = true;
    }

    private void TextBoxGotFocus(object? sender, GotFocusEventArgs e)
    {
        if (DataContext is not KeyboardShortcutInputViewModel vm) return;
        vm.StartListening();
    }

    private void TextBoxLostFocus(object? sender, RoutedEventArgs e)
    {
        if (DataContext is not KeyboardShortcutInputViewModel vm) return;
        vm.Cancel();
    }
}

File: src/Starter/Features/InternalSearchEngines/Settings/Views/KeyboardShortcutInput.axaml.fs
namespace Starter.Features.InternalSearchEngines.Settings.Views

open System
open Avalonia
open Avalonia.Controls
open Avalonia.Input
open Avalonia.Interactivity
open Avalonia.Markup.Xaml
open Starter.Features.InternalSearchEngines.Settings.ViewModels

type KeyboardShortcutInput() as this =
    inherit UserControl()

    let mutable dataContext = null
    let mutable subscription = ValueNone

    static let ControlToFocusProperty = AvaloniaProperty.Register<KeyboardShortcutInput, Control | null>("ControlToFocus", null)

    do this.InitializeComponent()

    member this.ControlToFocus
        with get () = this.GetValue(ControlToFocusProperty)
        and set v = this.SetValue(ControlToFocusProperty, v) |> ignore

    member this.InitializeComponent() =
        AvaloniaXamlLoader.Load this

    override this.Finalize() =
        subscription |> ValueOption.iter (fun d -> (d :> IDisposable).Dispose())

    override this.OnDataContextChanged _ =
        subscription |> ValueOption.iter (fun d -> (d :> IDisposable).Dispose())

        match this.DataContext with
        | :? KeyboardShortcutInputViewModel as dc ->
            dataContext <- dc
            subscription <-
                dc.StoppedListeningKeys.Subscribe(fun _ ->
                    match this.ControlToFocus with
                    | null -> ()
                    | c -> c.Focus() |> ignore
                )
                |> ValueSome
        | _ -> ()

    member this.TextBoxKeyDown(_sender: obj, args: KeyEventArgs) =
        match this.DataContext with
        | :? KeyboardShortcutInputViewModel as dc ->
            dc.KeyDown args.Key
            args.Handled <- true
        | _ -> ()

    member this.TextBoxGotFocus(_sender: obj, _args: RoutedEventArgs) =
        match this.DataContext with
        | :? KeyboardShortcutInputViewModel as dc -> dc.StartListeningKeys()
        | _ -> ()

    member this.TextBoxLostFocus(_sender: obj, _args: RoutedEventArgs) =
        match this.DataContext with
        | :? KeyboardShortcutInputViewModel as dc -> dc.Cancel()
        | _ -> ()
